mapstructure.go

```diff
@@ -862,6 +862,10 @@ func (d *Decoder) decodeMapFromStruct(name string, dataVal reflect.Value, val re
 		// Next get the actual value of this field and verify it is assignable
 		// to the map value.
 		v := dataVal.Field(i)
+		if v.Kind() == reflect.Ptr && v.Elem().Kind() == reflect.Struct {
+			// Handle embedded struct pointers as embedded structs.
+			v = v.Elem()
+		}
```

Comment:
I think this is what is causing the bug in #231 

If I remove this line my tests pass, but of course the new tested added in this PR fail.

I attempted to move this check to lower down, but I have not got it working yet. It seems like we should not be changing `v`, but we may need to deref the ptr in a few places to make this work.

Reply:
I got it to pass, one sec.

Reply:
https://github.com/mitchellh/mapstructure/pull/232